custom/icds/location_reassignment/processor.py

```diff
@@ -32,90 +23,11 @@ def __init__(self, domain, transitions):
-            new_locations_details = self.new_location_details.get(location_type_code, {})
-            for site_code, details in new_locations_details.items():
-                if details['parent_site_code']:
-                    existing_parent_site_codes.add(details['parent_site_code'])
-                # remove it from parent site codes if it itself needs to be created
-                existing_parent_site_codes.discard(site_code)
-
-        if existing_parent_site_codes:
-            return self._get_locations_by_site_codes(existing_parent_site_codes)
-        return {}
-
-    def _get_locations_by_site_codes(self, site_codes):
-        parents = SQLLocation.active_objects.filter(domain=self.domain, site_code__in=site_codes)
-        if len(parents) != len(site_codes):
-            missing_site_codes = site_codes - set([loc.site_code for loc in parents])
-            raise LocationCreateError(
-                "Could not find parent locations with following site codes %s" % ",".join(missing_site_codes))
-        return {
-            loc.site_code: loc for loc in parents
-        }
-
-    def _perform_transitions(self, operation, transitions):
-        # split operation has the old site code as the key whereas others have the new site code
-        for from_site_codes, to_site_codes in transitions.items():
-            if operation == SPLIT_OPERATION:
-                old_site_codes, new_site_codes = from_site_codes, to_site_codes
-            else:
-                new_site_codes, old_site_codes = from_site_codes, to_site_codes
-            errors = deprecate_locations(self.domain, self._get_locations(old_site_codes),
-                                         self._get_locations(new_site_codes), operation)
-            if errors:
-                raise InvalidTransitionError(",".join(errors))
-
-    def _get_locations(self, site_codes):
-        site_codes = site_codes if isinstance(site_codes, list) else [site_codes]
-        locations = [self.transiting_locations_by_site_code.get(site_code)
-                     for site_code in site_codes
-                     if self.transiting_locations_by_site_code.get(site_code)]
-        if len(locations) != len(site_codes):
-            loaded_site_codes = [loc.site_code for loc in locations]
-            missing_site_codes = set(site_codes) - set(loaded_site_codes)
-            raise InvalidTransitionError(
-                "Could not load location with following site codes: %s" % ",".join(missing_site_codes))
-        return locations
-
-    def update_users(self):
-        from custom.icds.location_reassignment.tasks import update_usercase
-        for old_username, new_username in self.user_transitions.items():
-            update_usercase.delay(self.domain, old_username, new_username)
+        # process each sheet, in order of hierarchy
+        # so that newly created parent locations are present when needed
+        for location_type_code in self.location_types_by_code:
```

Comment:
Since new locations are now created as a part of the transition itself, in case a location that is suppose to be created during the reassignment and is also used as a parent later, it needs to be created first. Hence we need to go from top to bottom.
I can't remember a strong reason why this was the other way around earlier, I guess we just wanted to archive all child locations before archiving their parent.
Instead of using the archive method on SQLLocation which archives a location and also its descendants, we are just marking the location as archived directly by setting is_archived to True. So this should not be a problem from archival perspective.

Looking at how HQ does archival, it does archive the parent location first and then its descendants, as seen [here](https://github.com/dimagi/commcare-hq/blob/6ff7128d1e39b35c513698f0238b3706297d2ef5/corehq/apps/locations/models.py#L547) where the query would return the parent location on top of the list, so i am thinking this will be fine.